src/pygeoprocessing/routing/routing.pyx

```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+            came from
+        discovery/finish (long): the discovery and finish time that defines
+            whether a pixel discovery time is inside a watershed or not.
+        discovery_managed_raster (_ManagedRaster): discovery time raster
+            x/y gives the discovery time for that pixel.
+        discovery_nodata (long): nodata value for discovery raster
+        boundary_list (list): this list is appended to for new pixels that
+            should be neighbors in the fill.
+
+    Return:
+        None.
+    """
+    # always add the current pixel
+    boundary_list.append((x_l, y_l))
+
+    # this section determines which back diagonal was in the watershed and
+    # fills it. if none are we pick one so there's no degenerate case
+    cdef int xdelta = D8_XOFFSET[edge_dir]
+    cdef int ydelta = D8_YOFFSET[edge_dir]
+    test_list = [
+        (x_l - xdelta, y_l),
+        (x_l, y_l - ydelta)]
+    for x_t, y_t in test_list:
+        point_discovery = <long>discovery_managed_raster.get(
+            x_t, y_t)
+        if (point_discovery != discovery_nodata and
+                point_discovery >= discovery and
+                point_discovery <= finish):
+            boundary_list.append((int(x_t), int(y_t)))
+            # there's only one diagonal to fill in so it's done here
+            return
+
+    # if there's a degenerate case then just add the xdelta,
+    # it doesn't matter
+    boundary_list.append(test_list[0])
+
+
+cdef int _in_watershed(
+        int x_l, int y_l, int direction_to_test, int discovery, int finish,
+        int n_cols, int n_rows,
+        _ManagedRaster discovery_managed_raster,
+        long discovery_nodata):
+    """Test if pixel in direction is in the watershed.
+
+    Args:
+        x_l/y_l (int): leading coordinate of the watershed boundary
+            edge.
+        direction_to_test (int): D8 direction that points which direction the edge
+            came from
+        discovery/finish (long): the discovery and finish time that defines
+            whether a pixel discovery time is inside a watershed or not.
+        n_cols/n_rows (int): number of columns/rows in the discovery raster,
+            used to ensure step does not go out of bounds.
+        discovery_managed_raster (_ManagedRaster): discovery time raster
+            x/y gives the discovery time for that pixel.
+        discovery_nodata (long): nodata value for discovery raster
+
+    Return:
+        1 if in, 0 if out.
+    """
+    cdef int x_n = x_l + D8_XOFFSET[direction_to_test]
+    cdef int y_n = y_l + D8_YOFFSET[direction_to_test]
+    if x_n < 0 or y_n < 0 or x_n >= n_cols or y_n >= n_rows:
+        return 0
+    cdef long point_discovery = <long>discovery_managed_raster.get(x_n, y_n)
+    return (point_discovery != discovery_nodata and
+            point_discovery >= discovery and
+            point_discovery <= finish)
+
+
+def _calculate_stream_geometry(

```

Comment:
Sure! changed in f97cfcd. I think I left it as a def because it returns a geometry object so there's overhead on that but it doesn't hurt anything. Mostly this is in a function because it's a complicated bit of code I needed to use a few different places and since we were already in Cython I static typed everything that made sense to.